packages/fxa-auth-server/lib/payments/stripe-firestore.ts

```diff
@@ -0,0 +1,211 @@
+      throw newFirestoreStripeError(
+        `Customer ${subscription.customer} was not found`,
+        FirestoreStripeError.FIRESTORE_CUSTOMER_NOT_FOUND
+      );
+    }
+
+    return customerSnap.docs[0].ref
+      .collection('subscriptions')
+      .doc(subscription.id)
+      .set(subscription, { merge: true });
+  }
+
+  /**
+   * Insert an invoice record into Firestore under the customer's stripe id.
+   */
+  async insertInvoiceRecord(invoice: Stripe.Invoice) {
+    const customerSnap = await this.customerCollectionDbRef
+      .where('id', '==', invoice.customer as string)
+      .get();
+    if (customerSnap.empty) {
+      throw newFirestoreStripeError(
+        `Customer ${invoice.customer} was not found`,
+        FirestoreStripeError.FIRESTORE_CUSTOMER_NOT_FOUND
+      );
+    }
+
+    if (typeof invoice.subscription !== 'string') {
+      throw new Error(`Invoice ${invoice.id} has no subscription id.`);
+    }
+
+    return customerSnap.docs[0].ref
+      .collection('subscriptions')
+      .doc(invoice.subscription)
+      .collection('invoices')
+      .doc(invoice.id)
+      .set(invoice, { merge: true });
+  }
+
+  /**
+   * Retrieve the customer from Firestore by either FxA uid or Stripe customer id.
+   */
+  async retrieveCustomer(
+    options:
+      | { uid: string; customerId: undefined }
+      | { uid: undefined; customerId: string }
+  ) {
+    if (options.uid) {
+      const customerSnap = await this.customerCollectionDbRef
+        .doc(options.uid)
+        .get();
+      if (customerSnap.exists) {
+        return customerSnap.data() as Stripe.Customer;
+      }
+    } else if (options.customerId) {
+      const customerSnap = await this.customerCollectionDbRef
+        .where('id', '==', options.customerId)
+        .get();
+      if (!customerSnap.empty) {
+        return customerSnap.docs[0].data() as Stripe.Customer;
+      }
+    }
+    throw newFirestoreStripeError(
+      `Customer ${options.customerId || options.uid} was not found`,
+      FirestoreStripeError.FIRESTORE_CUSTOMER_NOT_FOUND
+    );
+  }
+
+  /**
+   * Retrieve all the customer subscriptions from Firestore.
+   */
+  async retrieveCustomerSubscriptions(customerId: string) {
+    const customerSnap = await this.customerCollectionDbRef
+      .where('id', '==', customerId)
+      .get();
+    if (customerSnap.empty) {
+      throw newFirestoreStripeError(
+        `Customer ${customerId} was not found`,
+        FirestoreStripeError.FIRESTORE_CUSTOMER_NOT_FOUND
+      );
+    }
+
+    const subscriptionSnap = await customerSnap.docs[0].ref
+      .collection('subscriptions')
+      .get();
+    return subscriptionSnap.docs.map(
+      (doc) => doc.data() as Stripe.Subscription
+    );
+  }
+
+  /**
+   * Retrieve a subscription from Firestore by Stripe subscription id.
+   */
+  async retrieveSubscription(subscriptionId: string) {
+    const subscriptionSnap = await this.firestore
+      .collectionGroup('stripe-subscriptions')
```

Comment:
Just 'subscriptions'?

Reply:
Oh, sorry, I see the confusion. CollectionGroup names need to be unique for the entire GCP project. I was worried calling them just subscriptions or invoices means that name can't be used at the top level within this GCP project. They're like a special case of a global index.